### PR TITLE
Granularize toMemoryConfigOp

### PR DESCRIPTION
**Pull request by jnie-TT**
Introduce 5 ops that cover the functionality of the previous toMemoryConfigOp:
- ToDeviceOp: Move data to device
- FromDeviceOp: Move data from device to host
- ToLayoutOp: Tilize/untilize data
- TypecastOp: typecast data 
   - FYI @mmanzoorTT I updated the TTNN implementation of the typecast op, creating its own conversion pattern. In TTNN, typecast does not belong to eltwise unary, that's why you had to add a separate method for running the op in runtime. The stableHLO to ttir interface remains unchanged.
- ToMemoryConfigOp: Solely converts memory config, including Dram <-> L1 movement, interleaved <-> sharding conversion

Updated downstream consumers (TTNN to flatbuffer, TTNN Runtime, TTNNToEmitC)

Creating these ops in combination is a challenge, as there are several (known) constraints:
- Tilizing on device requires bf16
- Typecasting on device requires TILIZED data
- Untilizing on device requires width % 2 == 0, page size > sizeof(uint32_t).
  - Here it seems very hard to update the page size. Some ways I think might work is typecasting or padding. However padding -> untilize -> unpad seems tricky, and typecasting requires tilized data. 
  - Therefore for now we will always untilize on host (which is what we are doing currently), since we normally only untilize at the end of the program when we move data from device to host, the perf hit should be acceptable. We rarely have device to device untilize cases.

This PR also fixes tensor type when forcing row_major/tile layouts. Previously when forcing tile/row_major layout, the output tensor layout/shard shape/element type are not changed, and so runtime had to hack tilize/untilize logic instead of trusting the tile shape.

Also added a shardSpec attribute to the memoryConfigAttr because the optimizer needs it for reshard implementation.

Another goal here is to make adding multi-chip changes easier, as currently without this change we are missing a couple of ops, and also hacking toMemoryConfig ops in the compiler. I would thus need to update the compound toMemoryConfig op to support multichip, which will get removed anyway.

Here's a simple example to illustrate a couple issues and how they are fixed with this PR.
**On Main** we have
```
#layout = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<512x1024xbf16, #system>>
#layout1 = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<512x1024xbf16, #dram>, interleaved>
module attributes {tt.device = #device, tt.system_desc = #system_desc} {
  func.func @softmax(%arg0: tensor<512x1024xbf16, #layout>) -> tensor<512x1024xbf16, #layout> {
    %0 = "ttnn.get_device"() <{mesh_shape = #ttnn<mesh_shape 1x1>}> : () -> !tt.device<#device>
    %1 = "ttnn.to_layout"(%arg0, %0) <{layout = #ttnn.layout<tile>}> : (tensor<512x1024xbf16, #layout>, !tt.device<#device>) -> tensor<512x1024xbf16, #layout1>
    %2 = "ttnn.to_device"(%1, %0) <{memory_config = #ttnn.memory_config<<interleaved>, <dram>>}> : (tensor<512x1024xbf16, #layout1>, !tt.device<#device>) -> tensor<512x1024xbf16, #layout1>
    "ttnn.dealloc"(%1) : (tensor<512x1024xbf16, #layout1>) -> ()
    %3 = "ttnn.softmax"(%2) <{dimension = 1 : si32}> : (tensor<512x1024xbf16, #layout1>) -> tensor<512x1024xbf16, #layout1>
    "ttnn.dealloc"(%2) : (tensor<512x1024xbf16, #layout1>) -> ()
    %4 = "ttnn.softmax"(%3) <{dimension = -1 : si32}> : (tensor<512x1024xbf16, #layout1>) -> tensor<512x1024xbf16, #layout1>
    "ttnn.dealloc"(%3) : (tensor<512x1024xbf16, #layout1>) -> ()
    %5 = "ttnn.to_memory_config"(%4, %0) : (tensor<512x1024xbf16, #layout1>, !tt.device<#device>) -> tensor<512x1024xbf16, #layout>
    "ttnn.dealloc"(%4) : (tensor<512x1024xbf16, #layout1>) -> ()
    return %5 : tensor<512x1024xbf16, #layout>
  }
}
```
Notice we are hard-coding `to_layout` to tilize on host (to_layout always comes before to_device), and we are hacking in `ttnn.to_memory_config` at the end to move data from device to host and do all the conversions necessary at the end (typecasting data type, untilize, to_cpu etc.). There's no ability currently to typecast/convert memory config in the middle of the program - everything happens at the end when we're done.
Also notice the to_layout op that produces tensor %1: `%1 = "ttnn.to_layout"(%arg0, %0) <{layout = #ttnn.layout<tile>}> : (tensor<512x1024xbf16, #layout>, !tt.device<#device>) -> tensor<512x1024xbf16, #layout1>`. In the parameters of the op, we see `ttnn.layout<tile>`, which means this op is trying to tilize the tensor. However, notice the layout of the output tensor `#layout1... memref<512x1024xbf16, #dram>`, the element type is not tilized, instead it is still row_major. This causes a mismatch between op parameter and output tensor type, and further downstream, all consumers of this tensor will assume the tensor is not tilized when inspecting it's element type.

**With the fixes in this PR** we have
```
#layout = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<512x1024xbf16, #system>>
#layout1 = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<1x1x!tt.tile<32x32, bf16>, #dram>, interleaved>
#layout2 = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<512x1024xbf16, #dram>, interleaved>
module attributes {tt.device = #device, tt.system_desc = #system_desc} {
  func.func @softmax(%arg0: tensor<512x1024xbf16, #layout>) -> tensor<512x1024xbf16, #layout> {
    %0 = "ttnn.get_device"() <{mesh_shape = #ttnn<mesh_shape 1x1>}> : () -> !tt.device<#device>
    %1 = "ttnn.to_device"(%arg0, %0) <{memory_config = #ttnn.memory_config<<interleaved>, <dram>, <<1x1>>>}> : (tensor<512x1024xbf16, #layout>, !tt.device<#device>) -> tensor<512x1024xbf16, #layout1>
    %2 = "ttnn.to_layout"(%1) <{layout = #ttnn.layout<tile>}> : (tensor<512x1024xbf16, #layout1>) -> tensor<512x1024xbf16, #layout1>
    "ttnn.dealloc"(%1) : (tensor<512x1024xbf16, #layout1>) -> ()
    %3 = "ttnn.softmax"(%2) <{dimension = 1 : si32}> : (tensor<512x1024xbf16, #layout1>) -> tensor<512x1024xbf16, #layout2>
    "ttnn.dealloc"(%2) : (tensor<512x1024xbf16, #layout1>) -> ()
    %4 = "ttnn.softmax"(%3) <{dimension = -1 : si32}> : (tensor<512x1024xbf16, #layout2>) -> tensor<512x1024xbf16, #layout2>
    "ttnn.dealloc"(%3) : (tensor<512x1024xbf16, #layout2>) -> ()
    %5 = "ttnn.from_device"(%4) : (tensor<512x1024xbf16, #layout2>) -> tensor<512x1024xbf16, #layout>
    "ttnn.dealloc"(%4) : (tensor<512x1024xbf16, #layout2>) -> ()
    %6 = "ttnn.to_layout"(%5) <{layout = #ttnn.layout<row_major>}> : (tensor<512x1024xbf16, #layout>) -> tensor<512x1024xbf16, #layout>
    "ttnn.dealloc"(%5) : (tensor<512x1024xbf16, #layout>) -> ()
    return %6 : tensor<512x1024xbf16, #layout>
  }
}

```
Now, notice the first to_layout op is inserted after the to_device op, thus we are indeed tilizing on device now because the dataformat is bf16. Also notice there is no hacked to_memory_config in the end when the ops is finished. Instead, we call a from_device op that reads from device, and a to_layout op that untilizes on host, and finally return the tensor. Also notice the first to_layout_op: `%2 = "ttnn.to_layout"(%1) <{layout = #ttnn.layout<tile>}> : (tensor<512x1024xbf16, #layout1>) -> tensor<512x1024xbf16, #layout1>`. The output tensor layout is `#layout1`, which is now indeed in tile format: `memref<1x1x!tt.tile<32x32, bf16>`, which is aligned with the parameter/goal of the op. Now, all downstream consumers will see a tilized tensor.

Misc:
Synced with @svuckovicTT, we concluded that we should:
- Break down the big if statement into separate methods (Done, now all cases belong to their own function)
- Model to-layout op to include all parameters in TTNN as optional (Done in this PR)
- Convert typecast op to DPS, will create a separate issue for this, don't want to bloat this PR any further
- Create an issue that tracks `TensorMemoryLayout::None`. This won't be needed once we have TTNN-specific tensor type in our dialect, as we could just make the memory config optional

Additional changes: 
- Removed sharded softmax test because the outputs are wrong #843 
- Updated matmul to use ::ttnn::matmul
- Marked ttrt test_device as skipped because we are seeing perf issues on main: #828

### CODE REVIEW

**Comment by jnie-TT** [comment]
> Trying to understand this better. Is it true that TTNN to_device op can only move from host to dram, or can it also put the tensor directly into an L1 sharded configuration?

It can put it in a L1 sharded configuration directly, the to device op accepts a memory config as a parameter

File: test/ttmlir/Silicon/TTNN/sharded/simple_eltwise_sharded.mlir [hunk 1]
```diff
@@ -82,21 +82,8 @@ func.func @sqrt(%arg0: tensor<224x64xf32>) -> tensor<224x64xf32> {
...
-  // CHECK: %[[C:.*]] = "ttnn.softmax"[[C:.*]]
-  // Check for negative dimension attribute
-  %3 = "ttir.softmax"(%1, %2) <{dimension = -1 : si32, operand_constraints = [#l1_block_sharded, #l1_block_sharded]}> : (tensor<224x64xbf16>, tensor<224x64xbf16>) -> tensor<224x64xbf16>
-  return %3 : tensor<224x64xbf16>
-}
-
 /////////////////////////////////////////
 // Unsupported eltwise ops with sharding
 //  * Concat: Sharded concat requires ROW MAJOR layout
+//  * Softmax: Sharded softmax produces incorrect values
```
[the remark this answers is not in the document]
Comment (jnie-TT) [reply test/ttmlir/Silicon/TTNN/sharded/simple_eltwise_sharded.mlir:88]: AFAIK we need to pass in a specific parameter SoftmaxShardedMultiCoreProgramConfig for this to work with sharded data.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 2]
```diff
@@ -119,118 +131,579 @@ class TensorEmptyConversionPattern
...
-//    change the tensor to RowMajor layout, and then the TTNN::FromDeviceOp to
-//    move to host memory
-//
+// are required to achieve an arbitrary layout, namely:
+// - ToLayoutOp: to set the layout (ROW_MAJOR, TILE) of the tensor
+// - TypecastOp: to change the data type of the tensor
+// - ToDeviceOp: to move the tensor to a specific device
+// - FromDeviceOp: to move the tensor from a specific device to host
+// - ToMemoryConfigOp: to set the memory configuration (dram, l1, interleaved,
+// sharded) of the tensor
```
Comment (svuckovicTT) [inline lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:140]: Let's zoom about this, I'd like to understand what the intent here is.

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp [hunk 3]
```diff
@@ -376,7 +454,7 @@ class ToLayoutOpConversionPattern
     attrs.push_back(convertLayoutAttr(rewriter, srcOp.getLayoutAttr()));
     attrs.push_back(createStdNullopt(rewriter));
     attrs.push_back(createStdNullopt(rewriter));
-    attrs.push_back(mlir::IntegerAttr::get(rewriter.getIndexType(), 1));
+    attrs.push_back(createNullDevicePointer(rewriter));
```
Comment (svuckovicTT) [inline lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp:457]: Why this change?

**Review by svuckovicTT — CHANGES_REQUESTED** [review]
Hey @jnie-TT, I'm not sure I understand what the intent of this PR is. There's several changes that I think are not the way we want to go - I'd like to zoom over this to better understand what it is we're trying to accomplish here.

In general, we should strive to model what TTNN does today - those are the best practices that will save us a lot of headaches going forward.

**Comment by jnie-TT** [comment]
> Hey @jnie-TT, I'm not sure I understand what the intent of this PR is. There's several changes that I think are not the way we want to go - I'd like to zoom over this to better understand what it is we're trying to accomplish here.
> 
> In general, we should strive to model what TTNN does today - those are the best practices that will save us a lot of headaches going forward.

@svuckovicTT this essentially builds on top of the changes you added in #624 that replaces toMemoryConfigOp with more granular ops. However there were some missing ops (typecast, toMemoryConfig), and the update did not check relevant constraints, so a partial revert was needed because fromDeviceOp was not working properly. 

With this we now replace the old compound toMemoryConfigOp with multiple ops with more granular control, while satisfying ttnn constraints.

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td [hunk 4]
```diff
@@ -42,17 +49,34 @@ def TTNN_ToMemoryConfigOp : TTNN_Op<"to_memory_config"> {
 def TTNN_ToLayoutOp : TTNN_Op<"to_layout"> {
     let summary = "ToLayout op.";
     let description = [{
+      This op converts the layout of the input tensor based on the given layout.
+      It handles:
+        - ROW_MAJOR to TILE
+        - TILE to ROW_MAJOR
```
[the remark this answers is not in the document]
Comment (jnie-TT) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:55]: a toLayoutOp will not be created if it's ROW_MAJOR -> ROW_MAJOR, only when the input/output layouts are different will a toLayoutOp be created

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td [hunk 5]
```diff
@@ -42,17 +49,34 @@ def TTNN_ToMemoryConfigOp : TTNN_Op<"to_memory_config"> {
...
     }];
 
     let arguments = (ins AnyRankedTensor:$input,
-                         TT_Device:$device,
```
[the remark this answers is not in the document]
Comment (jnie-TT) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:48]: When moving a tensor to device we should use the toDeviceOp, toLayoutOp should be solely used to tilize/untilize tensors. If there's too much overlap between functionality of ops then it makes it hard to model in the compiler, and these boundaries should be set between ops.

File: include/ttmlir/Dialect/TTNN/IR/TTNNOpsEnums.td [hunk 6]
```diff
@@ -21,14 +21,16 @@ def TTNN_Layout : I32EnumAttr<"Layout", "TTNN Layout",
   let cppNamespace = "::mlir::tt::ttnn";
 }
 
-def TTNN_TensorMemoryLayout_Interleaved : I32EnumAttrCase<"Interleaved", 0, "interleaved">;
-def TTNN_TensorMemoryLayout_SingleBank : I32EnumAttrCase<"SingleBank", 1, "single_bank">;
-def TTNN_TensorMemoryLayout_HeightSharded : I32EnumAttrCase<"HeightSharded", 2, "height_sharded">;
-def TTNN_TensorMemoryLayout_WidthSharded : I32EnumAttrCase<"WidthSharded", 3, "width_sharded">;
-def TTNN_TensorMemoryLayout_BlockSharded : I32EnumAttrCase<"BlockSharded", 4, "block_sharded">;
+def TTNN_TensorMemoryLayout_None : I32EnumAttrCase<"None", 0, "none">;
```
[the remark this answers is not in the document]
Comment (jnie-TT) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOpsEnums.td:24]: With this all host tensors should have None tensor memory layout. It wouldn't make sense for host tensors to have other memory layouts. If you look at the get_memory_config API in ttnn, they model host tensors with std::nullopt tensor memory layouts (wrapped in the memory config). This mimics that behaviour:
```
inline std::optional<ttnn::MemoryConfig> get_memory_config(const ttnn::Tensor& tensor) {
    if (not tensor.is_allocated() or not is_tensor_on_device_or_multidevice(tensor)) {
        return std::nullopt;
    }
    return tensor.memory_config();
}
```

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp (hunk 3 above)
Comment (jnie-TT) [reply lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp:457]: Because toLayoutOps shouldn't be moving tensors to device, that's the responsibility of toDeviceOps. Therefore we set the device pointer to null.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp (hunk 2 above)
Comment (jnie-TT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:140]: Sure I'd be happy to zoom. In the meantime hopefully this [comment](https://github.com/tenstorrent/tt-mlir/pull/840#issuecomment-2387283194) clarifies things a bit.

**Comment by jnie-TT** [comment]
> Hey @jnie-TT, I'm not sure I understand what the intent of this PR is. There's several changes that I think are not the way we want to go - I'd like to zoom over this to better understand what it is we're trying to accomplish here.
> 
> In general, we should strive to model what TTNN does today - those are the best practices that will save us a lot of headaches going forward.

@svuckovicTT this essentially builds on top of the changes you added in #624 that replaces toMemoryConfigOp with more granular ops. 

To summarize the current state/issues of memory ops are:
- We're missing a couple ops (typecast, toMemoryConfig). Here toMemoryConfig solely converts the memory config of the tensor.
- We're always tilizing on host (toLayout ops are always inserted before toDeviceOps). I was told by the TTNN folks that the host tilizer is very bad, and we should always tilize on device if possible.
- We're always assuming toLayout should move tensors to device (because we're always passing in the device argument). Therefore we cannot untilize on host (which is fixed by the next bullet).
- We're hacking in the old compound toMemoryConfigOp when reading tensors from device because there are issues with toLayout and fromDevice ops (#789 ).
- We have a compound toMemoryConfig op that checks everything, which eliminates granular control and is hard to update for specific use-cases. This op should solely convert memory configs of the device tensor.
- toDeviceOps are created when tensors are already on device (no checks are in place), related PR: #819 
- toLayoutOps are created when input output tensors could already have the same layout (no checks in place)

This change attempts to fix the above issues:
- Added typecast and toMemoryConfig ops
- We determine where to tilize (if the tensor can be tilized on device and the tensor will be moved to device/is on device, we tilize on device)
- We only move tensors to device with toDeviceOps, clear boundaries are set
- We use fromDevice ops to read tensors back from device, and we always untilize on host (because device untilize has many issues mentioned in the PR description)
- We remove the compound toMemoryConfig op, and make it lightweight such that all it does it convert memory configs of device tensors
- We check whether toDeviceOps should be created, and abide with constraints (e.g. should not be created in conjunction with fromDeviceOps, should not be created if tensor is already on device, should not be created before an illegal tilize/typecast etc.)
- We check the input/output tensor layout and determine if a toLayoutOp should be created

The way I see it, these issues should be solved eventually, and should be solved before any further multi-chip changes go in. Or else we would need to update a fromDeviceOp that's not being used (therefore cannot be tested easily), a compound toMemoryConfigOp that's checking everything (therefore tons of updates will need to go in which will probably get removed anyway when we granularize it), and a toDeviceOp + toLayoutOp that has mixed boundaries and are also not being properly checked.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp (hunk 2 above)
Comment (svuckovicTT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:140]: I don't want us to granularize this way for reasons mentioned in one of the comments above - we should model the `from_torch` approach, which is what the current code tries to do. Additionally, this piece of the change introduces a lot more code, which I find hard to read and maintain.

Can you post a list of specific scenarios we want to support, but are not currently supported?

File: runtime/lib/ttnn/operations/matmul/matmul.cpp [hunk 7]
```diff
@@ -15,11 +15,25 @@ void run(const ::tt::target::ttnn::MatmulOp *op, ProgramContext &context) {
...
+  const std::optional<const ::ttnn::operations::matmul::MatmulProgramConfig>
+      programConfig = std::nullopt;
+  const std::optional<const std::string> activation = std::nullopt;
+  const std::optional<const ::ttnn::DeviceComputeKernelConfig>
+      computeKernelConfig = std::nullopt;
+  const std::optional<const ::ttnn::CoreGrid> coreGrid = std::nullopt;
+
+  ::ttnn::Tensor out =
+      ::ttnn::matmul(lhs, rhs, transposeA, transposeB, memoryConfig, dtype,
+                     programConfig, activation, computeKernelConfig, coreGrid);
```
Comment (svuckovicTT) [inline runtime/lib/ttnn/operations/matmul/matmul.cpp:35]: If we need to do this for matmul (let's first prove that we do), we should reflect it in the dialect itself. That way, EmitC can work as well.

File: test/ttmlir/Silicon/TTNN/sharded/simple_eltwise_sharded.mlir (hunk 1 above)
[the remark this answers is not in the document]
Comment (svuckovicTT) [reply test/ttmlir/Silicon/TTNN/sharded/simple_eltwise_sharded.mlir:88]: Can you add an issue to our gitlab to track this, with all the details, and put a TODO with bug id in the comment?

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td [hunk 8]
```diff
@@ -42,17 +49,34 @@ def TTNN_ToMemoryConfigOp : TTNN_Op<"to_memory_config"> {
...
                          TTNN_LayoutAttr:$layout);
     let results = (outs AnyRankedTensor:$result);
 }
 
+def TTNN_TypecastOp : TTNN_Op<"typecast"> {
```
Comment (svuckovicTT) [inline include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:63]: `ttnn::experimental::typecast` op is a DPS op, we should adjust for that.

File: lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp [hunk 9]
```diff
@@ -355,6 +360,79 @@ class FromDeviceOpConversionPattern
   }
 };
 
+// TypecastOp conversion pattern
+//
+class TypecastOpConversionPattern
```
Comment (svuckovicTT) [inline lib/Conversion/TTNNToEmitC/TTNNToEmitC.cpp:365]: The op lives in `ttnn::experimental` namespace, you can override the `getPrefixSwapPattern()`. `GetDeviceOpConversionPattern` has an example.

**Review by svuckovicTT — CHANGES_REQUESTED** [review]

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 10]
```diff
@@ -119,21 +119,85 @@ class TensorEmptyConversionPattern
...
 class ToLayoutOpConversionPattern
     : public OpConversionPattern<ttir::ToLayoutOp> {
 public:
   using OpConversionPattern<ttir::ToLayoutOp>::OpConversionPattern;
 
+  LogicalResult
+  matchAndRewrite(ttir::ToLayoutOp op, OpAdaptor adaptor,
```
Comment (nobradovictt) [inline lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:135]: My ask here would be to try and migrate this decomposition of ToLayoutOp to a separate TTNN pass which occurs after lowering and thus keep lowering simple and clean. This would also allow optimizer to work with single memory layout manipulation op(for example ttnn:ToLayoutOp) while a separate pass would properly decompose it.

Comment (jnie-TT) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:129]: Added a separate pass after the optimizer that decomposes the composite toLayout op instead of doing the decomposing during the lowering stage. Updated optimizer accordingly

Comment (nobradovictt) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:129]: @jnie-TT Looks great on a first pass, will take a more detailed look on Monday morning.

**Review by tapspatel — APPROVED** [review]

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td [hunk 11]
```diff
@@ -27,13 +27,36 @@ def TTNN_GetDeviceOp : TTNN_Op<"get_device"> {
     let results = (outs TT_Device:$device);
 }
 
+def TTNN_CompositeToLayoutOp : TTNN_Op<"composite_to_layout"> {
```
Comment (nsmithtt) [inline include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:30]: Is this going to be permanent?  Perhaps we can replace this with ToLayout? As I understand it TTNN folks plan to make `ttnn::to_layout` flexible enough to support all kinds of layout transitions.  In the meantime we can specialize ToLayout to the respective lowered op.

**Review by nsmithtt — APPROVED** [review]

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td (hunk 11 above)
Comment (jnie-TT) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:30]: @nsmithtt that's a good point. I guess I can rename this op to ToLayout, and then the downstream decomposed ToLayout to something like TTNN_TilizeUntilize. Later when `ttnn::to_layout` supports all combinations of parameters, we can simply remove the downstream decomposition pass and use this op instead.

Comment (nobradovictt) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:30]: Yeah if you are going to change the name it's good idea to change downstream decomposed to something else.

File: lib/Dialect/TTNN/Analysis/DFShardingPolicy.cpp [hunk 12]
```diff
@@ -8,9 +8,8 @@
 
 namespace mlir::tt::ttnn {
 
-bool isMemoryManagementOp(mlir::Operation *op) {
-  return isa<ttnn::ToLayoutOp>(op) || isa<ttnn::ToMemoryConfigOp>(op) ||
-         isa<ttnn::ToDeviceOp>(op);
+bool isCompositeToLayoutOp(mlir::Operation *op) {
+  return isa<ttnn::CompositeToLayoutOp>(op);
 }
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Analysis/DFShardingPolicy.cpp:13]: No need for this util function anymore since it is a single op again.

File: lib/Dialect/TTNN/Pipelines/TTNNPipelines.cpp [hunk 13]
```diff
@@ -98,7 +121,7 @@ void registerTTNNPipelines() {
   mlir::PassPipelineRegistration<
       mlir::tt::ttnn::TTIRToTTNNBackendPipelineOptions>(
       "ttir-to-ttnn-backend-pipeline",
-      "Pipeline lowering ttir to ttmetal backend.",
+      "Pipeline lowering ttir to ttnn backend.",
```
Comment (nobradovictt) [inline lib/Dialect/TTNN/Pipelines/TTNNPipelines.cpp:124]: Nice catch :)

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp (hunk 10 above)
Comment (nobradovictt) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:129]: Looks good, please proceed.

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td (hunk 11 above)
Comment (nsmithtt) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:30]: Or we just leave it?  At the initial conversion / optimizer just targets:
- `ToLayout`

Then, in the short term, we might have to lower `ToLayout` to one of:
- `ToMemoryConfig`
- `Typecast`
- etc.
- Or just leave it, if it's supported, as `ToLayout`

Over time, we can just erase the specialized lowering for cases where `ToLayout` has added support for it.

Comment (nobradovictt) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:30]: I don't have a hard preference here but leaving the same name is in my opinion bug prone. Reordering passes may create hard to debug issues with Op which has different meaning depending on place it's used.

Comment (nsmithtt) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:30]: @jnie-TT yeah that's right.

@nobradovictt, I agree it's not ideal, but neither is introducing a dummy op esp if the eventual semantics of the `ToLayout` op are exactly how we want to use it.  Think of it as a workaround, we lower to `ToLayout`, workaround passes at the end get applied that specialize out all the cases that `ToLayout` doesn't support yet.

**Review by svuckovicTT — APPROVED** [review]
Hey @jnie-TT, thanks for taking the time to rework the ToLayout conversion, I think this is much cleaner now :)

Looks good from dialects perspective, can you please make sure to add the missing issues? (layout=none, typecast to dps, and dtype=none)

Great work!

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td (hunk 11 above)
Comment (nobradovictt) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:30]: @svuckovicTT I could make the same argument against having an op of the same name as in TTNN lib which carries completely different meaning(that it pushes TTNN dialect further from TTNN lib). I personally actually find it even worse, it's "wrong result" but insidious. It's easy to assert that dummy op is not expected in final IR.
But as this is a temp workaround, and more of you seems to prefer same name option, Im fine with that.

**Comment by jnie-TT** [comment]
> Hey @jnie-TT, thanks for taking the time to rework the ToLayout conversion, I think this is much cleaner now :)
> 
> Looks good from dialects perspective, can you please make sure to add the missing issues? (layout=none, typecast to dps, and dtype=none)
> 
> Great work!

For sure, I'll create separate issues for this to track. Thanks!

File: include/ttmlir/Dialect/TTNN/IR/TTNNOps.td (hunk 11 above)
Comment (jnie-TT) [reply include/ttmlir/Dialect/TTNN/IR/TTNNOps.td:30]: I'll create a separate issue to track this.

**Review by AleksKnezevic — APPROVED** [review]
